Add thank you page to Jetpack Userless Checkout

cbauerman opened the pull request:
#### Changes proposed in this Pull Request

* Add new thank-you screen exclusively for Jetpack ( "userless" ) Checkout purchases
* Modify composite checkout to use the jetpack site slug in all cases in the event of a Jetpack "userless" checkout
* Add a new shortcut `getProductName` selector.

![Untitled 12](https://user-images.githubusercontent.com/2810519/117895061-7abaa400-b272-11eb-8e43-358bad624373.png)
![Untitled 4](https://user-images.githubusercontent.com/2810519/117895072-7db59480-b272-11eb-8930-7032dd41c50c.png)


#### Testing instructions

1. Complete a test purchase with instructions for parent PR #51953
2. Verify that one purchase completion you are redirected to the thank-you page ( `/checkout/jetpack/thank-you` )
3. Verify the design matches the various examples above

cbauerman committed:
Add basic thank-you component

cbauerman committed:
Add new request unauthorized site data getter

cbauerman committed:
Add product info and full translated strings

cbauerman committed:
Update `JetpackCheckout` variable name

cbauerman committed:
Rename component, move data getters, more styling

rcanepa committed:
Make 'onboarding-registrationless' default

cbauerman committed:
Rename controller method to new standard

cbauerman committed:
Supress masterbar

cbauerman committed:
Desktop layout style refinements

a8ci18n commented on client/my-sites/checkout/checkout-thank-you/jetpack-checkout-thank-you.tsx:
```diff
@@ -0,0 +1,116 @@
+					primary
+				>
+					{ isLoading
+						? translate( 'loading' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **84** times:
`translate( 'Loading' )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/try/jetpack-checkout-thank-you)

cbauerman committed:
Mobile style refinements

cbauerman committed:
Update loading button copy and style

cbauerman commented on client/layout/logged-out.jsx:
```diff
@@ -54,6 +54,10 @@ const LayoutLoggedOut = ( {
 	const isJetpackCheckout =
 		sectionName === 'checkout' && window.location.pathname.startsWith( '/checkout/jetpack' );
 
+	const isJetpackThankYou =
```
Is this a good practice? There doesn't seem to be another way to prevent the master bar from rendering.

cbauerman committed:
Initial flow setup

# Conflicts:
#	client/my-sites/checkout/composite-checkout/hooks/use-create-payment-complete-callback.tsx
#	client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts
#	client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/index.tsx

cbauerman committed:
Update composite to use jetpack site slug

# Conflicts:
#	client/my-sites/checkout/composite-checkout/hooks/use-create-payment-complete-callback.tsx

cbauerman committed:
Import changes from base branch

cbauerman committed:
Import changes from trunk

cbauerman committed:
Remove extra paren

sirbrillig commented on client/my-sites/checkout/composite-checkout/composite-checkout.tsx:
```diff
@@ -211,7 +211,7 @@ export default function CompositeCheckout( {
 		isJetpackNotAtomic,
 		isPrivate,
-		siteSlug,
+		siteSlug: isJetpackCheckout ? jetpackSiteSlug : siteSlug,
```
This pattern happens in a lot of places. I wonder if it would be worth hoisting it up a bit? That way we don't risk missing somewhere (especially in future commits). For example, at the top of this component we could write:

```js
siteSlug = isJetpackCheckout ? jetpackSiteSlug : siteSlug;
```

Then all of these instances could remain as they are. Unless there's some place in this component where we want the original `siteSlug` (I haven't looked yet).

sirbrillig commented on client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:
```diff
@@ -123,6 +125,16 @@ export default function getThankYouPageUrl( {
+		// extract a product from the cart, in userless checkout there should only be one
+		const productSlug = cart?.products[ 0 ]?.product_slug;
+
+		return `/checkout/jetpack/thank-you/${ siteSlug }/${ productSlug }`;
```
This function is covered by tests in [this file](https://github.com/Automattic/wp-calypso/blob/126ca2963ada08735986e411900ee1021a69594a/client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js#L52). Could we add a test for this case?

sirbrillig commented on client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:
```diff
@@ -123,6 +125,16 @@ export default function getThankYouPageUrl( {
+		debug( 'redirecting to userless jetpack thank you' );
+
+		// extract a product from the cart, in userless checkout there should only be one
+		const productSlug = cart?.products[ 0 ]?.product_slug;
```
It's very unlikely, but since this expression could evaluate to `undefined`, should we provide a fallback behavior so that the user isn't redirected to `/checkout/jetpack/thank-you/example.com/undefined`?

Two options I can think of:

Provide an empty string as a fallback: 

```const productSlug = cart?.products[ 0 ]?.product_slug ?? '';```

Or, skip this behavior:

```
if ( productSlug ) {
  return `/checkout/jetpack/thank-you/${ siteSlug }/${ productSlug }`;
}
```

cbauerman replied:
I built the page at the end of the route to handle not getting a product slug in case anything went wrong, so I think this would be the elegant approach to make sure we stay in the flow

```
return productSlug ? `/checkout/jetpack/thank-you/${ siteSlug }/${ productSlug }` : `/checkout/jetpack/thank-you/${ siteSlug }`;
```

cbauerman replied:
with the `/${ productSlug }` on the route it ended up being caught by another handler, so I opted to use a `no_product` for the product slug if an actual one cannot be found. The component handles `no_product` as a special case.

sirbrillig commented on client/my-sites/checkout/upsell-nudge/purchase-modal/index.tsx:
```diff
@@ -77,6 +77,7 @@ export function PurchaseModal( {
 export default function PurchaseModalWrapper( props: PurchaseModalProps ): JSX.Element {
 	const onComplete = useCreatePaymentCompleteCallback( {
 		isComingFromUpsell: true,
+		siteSlug: props.siteSlug,
```
Thank you for remembering this file! It's easy to overlook.

cbauerman committed:
Handle not finding product in getThankYouPageUrl

cbauerman committed:
Add tests for jetpackCheckout thank you url

cbauerman committed:
Consolidate jetpackSiteSlug logic in checkout

cbauerman committed:
Use const for updated SiteId in checkout

cbauerman committed:
Remove unecessary arguement reassignment
